src/payload.rs

```diff
@@ -7,9 +8,140 @@ use core::ops::{Deref, DerefMut};
 // | SW USE                        |               ACTUAL DMA PART                                    |
 // | rssi - 1 byte | pipe - 1 byte | length - 1 byte | pid_no_ack - 1 byte | payload - 1 to 252 bytes |
 
+/// A builder for an `EsbHeader` structure
+///
+/// The builder is converted into an `EsbHeader` by calling the
+/// `check()` method.
+///
+/// ## Example
+///
+/// ```rust
+/// use esb::EsbHeaderBuilder;
+///
+/// let header_result = EsbHeaderBuilder::default()
+///     .max_payload(252)
+///     .pid(0)
+///     .pipe(0)
+///     .no_ack(true)
+///     .check();
+///
+/// assert!(header_result.is_ok());
+/// ```
+///
+/// ## Default Header Contents
+///
+/// By default, the following settings will be used:
+///
+/// | Field     | Default Value |
+/// | :---      | :---          |
+/// | pid       | 0             |
+/// | no_ack    | false         |
+/// | length    | 0             |
+/// | pipe      | 0             |
+///
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct EsbHeaderBuilder(EsbHeader);
+
+impl Default for EsbHeaderBuilder {
+    fn default() -> Self {
+        EsbHeaderBuilder(EsbHeader {
+            rssi: 0,
+            pid_no_ack: 0,
+            length: 0,
+            pipe: 0,
+        })
+    }
+}
+
+impl EsbHeaderBuilder {
+    /// Set the pipe. Must be in the range 0..=7.
+    pub fn pipe(mut self, pipe: u8) -> Self {
+        self.0.pipe = pipe;
+        self
+    }
+
+    /// Set the max payload. Must be in the range 0..=252.
+    pub fn max_payload(mut self, max_payload: u8) -> Self {
+        self.0.length = max_payload;
+        self
+    }
+
+    /// Enable/disable acknowledgment
+    pub fn no_ack(mut self, no_ack: bool) -> Self {
+        // TODO(AJM): We should probably just call this
+        // method "ack", or "enable_ack", because "no_ack"
+        // is really confusing.
+        if no_ack {
+            self.0.pid_no_ack &= 0b1111_1110;
+        } else {
+            self.0.pid_no_ack |= 0b0000_0001;
+        }
+        self
+    }
+
+    /// Set the pid. Must be in the range 0..=3.
+    pub fn pid(mut self, pid: u8) -> Self {
+        // TODO(AJM): Do we want the user to set the pid? isn't this an
+        // internal "retry" counter?
```

Comment:
It's used for identifying repeated packages (together with CRC), I also feel that we should take care of it ourselves, the downside is that we would have to keep track of the pids of every pipe, Nordic leaves this to the user. I guess we can leave like this for now and once it's working we can come back to this, what do you think ?

Reply:
No objections! Happy to leave as-is, and we can clean up all the TODOs before the final release.